lib/elixir/test/elixir/kernel/diagnostics_test.exs

```diff
@@ -10,6 +10,305 @@ defmodule Kernel.DiagnosticsTest do
+          error: unexpected token: }
+          HINT: the "[" on line 1 is missing terminator "]"
+          │
+        1 │ [a, b, c
+          │ └ unclosed delimiter
+        2 │  d, f, g}
+          │         └ mismatched closing delimiter
+          │
+          └─ nofile:2:9\
+      """
+
+      output =
+        capture_raise(
+          """
+          [a, b, c
+           d, f, g}
+          """,
+          MismatchedDelimiterError
+        )
+
+      assert output == expected
+    end
+
+    test "many-line span" do
+      expected = """
+      ** (MismatchedDelimiterError) mismatched delimiter found on nofile:5:5:
+          error: unexpected token: )
+          HINT: the "[" on line 1 is missing terminator "]"
+          │
+        1 │ [ a,
+          │ └ unclosed delimiter
+        2 │   b,
+        3 │   c,
+        4 │   d
+        5 │   e )
+          │     └ mismatched closing delimiter
+          │
+          └─ nofile:5:5\
+      """
+
+      output =
+        capture_raise(
+          """
+          [ a,
+            b,
+            c,
+            d
+            e )
+          """,
+          MismatchedDelimiterError
+        )
+
+      assert output == expected
+    end
+
+    test "trim inbetween lines if too many" do
+      expected = """
+      ** (MismatchedDelimiterError) mismatched delimiter found on nofile:9:5:
+          error: unexpected token: )
+          HINT: the "[" on line 1 is missing terminator "]"
+          │
+        1 │ [ a,
+          │ └ unclosed delimiter
+          │ ...
+        9 │   i )
+          │     └ mismatched closing delimiter
+          │
+          └─ nofile:9:5\
+      """
+
+      output =
+        capture_raise(
+          """
+          [ a,
+            b,
+            c,
+            d,
+            e,
+            f,
+            g,
+            h
+            i )
+          """,
+          MismatchedDelimiterError
+        )
+
+      assert output == expected
+    end
+
+    test "pads according to line number digits" do
+      expected = """
+      ** (MismatchedDelimiterError) mismatched delimiter found on nofile:13:5:
+          error: unexpected token: )
+          HINT: the "[" on line 1 is missing terminator "]"
+          │
+        1 │ [ a,
+          │ └ unclosed delimiter
+          │ ...
+       13 │   b )
+          │     └ mismatched closing delimiter
+          │
+          └─ nofile:13:5\
+      """
+
+      output =
+        capture_raise(
+          """
+          [ a,
+          #{String.duplicate("\n", 10)}
+            b )
+          """,
+          MismatchedDelimiterError
+        )
+
+      assert output == expected
+
+      expected = """
+      ** (MismatchedDelimiterError) mismatched delimiter found on nofile:403:5:
+           error: unexpected token: )
+           HINT: the "[" on line 1 is missing terminator "]"
+           │
+         1 │ [ a,
+           │ └ unclosed delimiter
```

Comment:
Do you think it is worth showing one line before the opening and one line before the closing? Or would that be just additional noise? It may be helpful for the closing one because often it is just the closing token on its own line.

Reply:
I think that's valid, especially for the closing one, I'll give it a try and see how it looks like

Reply:
It's currently looking like this, wdyt?

### Code

```ex
MyLongModuleWithManyFeatures.call(fn a, b, c ->
  call_svc_1(:a)
  call_svc_2(:b)
  call_svc_3(:c)
  call_svc_4(:d)
  call_svc_5(:e)
)
```

### Error
```
** (MismatchedDelimiterError) mismatched delimiter found on unclosed.ex:7:1:
    error: unexpected token: )
    HINT: the "fn" on line 1 is missing terminator "end"
    │
  1 │ MyLongModuleWithManyFeatures.call(fn a, b, c ->
    │                                   └ unclosed delimiter
    │ ...
  6 │   call_svc_5(:e)
  7 │ )
    │ └ mismatched closing delimiter
    │
    └─ unclosed.ex:7:1
    (elixir 1.16.0-dev) lib/code.ex:1463: Code.require_file/2

```

Reply:
Good point about the line numbers, we can try without adding the extra line and see where it leads us.